drivers/timer/arm_arch_timer.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2019 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/timer/arm_arch_timer.h>
+#include <drivers/timer/system_timer.h>
+#include <sys_clock.h>
+#include <spinlock.h>
+#include <arch/cpu.h>
+
+#define CYC_PER_TICK	((u32_t)((u64_t)sys_clock_hw_cycles_per_sec() \
+			       / (u64_t)CONFIG_SYS_CLOCK_TICKS_PER_SEC))
+
+#define MAX_TICKS	((0xffffffffu - CYC_PER_TICK) / CYC_PER_TICK)
+#define MIN_DELAY	(1000)
+
+static struct k_spinlock lock;
+static volatile u64_t last_cycle;
+
+static void arm_arch_timer_compare_isr(void *arg)
+{
+	ARG_UNUSED(arg);
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	u64_t curr_cycle = arm_arch_timer_count();
+	u32_t delta_ticks = (u32_t)((curr_cycle - last_cycle) / CYC_PER_TICK);
+
+	last_cycle += delta_ticks * CYC_PER_TICK;
+
+	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL) ||
+	     IS_ENABLED(CONFIG_QEMU_TICKLESS_WORKAROUND)) {
+		u64_t next_cycle = last_cycle + CYC_PER_TICK;
+
+		if ((s64_t)(next_cycle - curr_cycle) < MIN_DELAY) {
+			next_cycle += CYC_PER_TICK;
+		}
+		arm_arch_timer_set_compare(next_cycle);
+	}
+
+	k_spin_unlock(&lock, key);
+
+	z_clock_announce(IS_ENABLED(CONFIG_TICKLESS_KERNEL) ? delta_ticks : 1);
+}
+
+int z_clock_driver_init(struct device *device)
+{
+	ARG_UNUSED(device);
+
+	IRQ_CONNECT(ARM_ARCH_TIMER_IRQ, 0, arm_arch_timer_compare_isr, 0,
+		    ARM_TIMER_FLAGS);
+	arm_arch_timer_set_compare(arm_arch_timer_count() + CYC_PER_TICK);
+	arm_arch_timer_toggle(true);
```

Comment:
`arm_arch_timer_enable()`  will be better ?

Reply:
ACK. I will rename it in the next revision.